pyccel/parser/syntactic.py

```diff
@@ -633,54 +644,126 @@ def _visit_FunctionDef(self, stmt):
+        # extract the templates
+        if 'template' in decorators:
+            if not isinstance(decorators['template'], list):
+                decorators['template'] = [decorators['template']]
+            for comb_types in decorators['template']:
+                cache.clear_cache()
+                types = []
+                if len(comb_types.args) != 2:
+                    msg = 'Number of Arguments provided to the template decorator is not valid'
+                    errors.report(msg,
+                                    symbol = comb_types,
+                                    bounding_box = (stmt.lineno, stmt.col_offset),
+                                    severity='error')
+
+                for i in comb_types.args:
+                    if isinstance(i, ValuedArgument) and not i.name in ('name',
+                            'types'):
+                        msg = 'Argument provided to the template decorator is not valid'
                         errors.report(msg,
-                                      symbol = decorators['types'],
-                                      bounding_box = (stmt.lineno, stmt.col_offset),
-                                      severity='error')
-                    else:
-                        ls = arg if isinstance(arg, PythonTuple) else [arg]
-                        i = -1
+                                        symbol = comb_types,
+                                        bounding_box = (stmt.lineno, stmt.col_offset),
+                                        severity='error')
+                if all(isinstance(i, ValuedArgument) for i in comb_types.args):
+                    tp_name, ls = (comb_types.args[0].value, comb_types.args[1].value) if\
+                            comb_types.args[0].name == 'name' else\
+                            (comb_types.args[1].value, comb_types.args[0].value)
                 else:
-                    msg = 'Invalid argument of type {} passed to types decorator'.format(type(arg))
+                    tp_name = comb_types.args[0]
+                    ls = comb_types.args[1]
+                    ls = ls.value if isinstance(ls, ValuedArgument) else ls
+                try:
+                    tp_name = str(tp_name)
+                    ls = ls if isinstance(ls, PythonTuple) else list(ls)
+                except TypeError:
+                    msg = 'Argument provided to the template decorator is not valid'
                     errors.report(msg,
-                                  symbol = decorators['types'],
-                                  bounding_box = (stmt.lineno, stmt.col_offset),
-                                  severity='error')
+                                    symbol = comb_types,
+                                    bounding_box = (stmt.lineno, stmt.col_offset),
+                                    severity='fatal')
+
+                i = 0
```

Comment:
Unused variable?

Reply:
yes, it didn't get detected, I just removed it

Reply:
Probably there is a variable with the same name in the same scope which is used